fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryIndexPlan.java

```diff
@@ -84,37 +100,103 @@ public class RecordQueryIndexPlan implements RecordQueryPlanWithNoChildren, Reco
-        this(indexName, scanParameters, reverse, strictlySorted, Optional.empty(), new Type.Any());
+        this(indexName, commonPrimaryKey, scanParameters, useIndexPrefetch, reverse, strictlySorted, Optional.empty(), new Type.Any());
     }
 
     public RecordQueryIndexPlan(@Nonnull final String indexName,
+                                @Nullable final KeyExpression commonPrimaryKey,
                                 @Nonnull final IndexScanParameters scanParameters,
+                                @Nonnull final RecordQueryPlannerConfiguration.IndexPrefetchUse useIndexPrefetch,
                                 final boolean reverse,
                                 final boolean strictlySorted,
                                 @Nonnull final ScanWithFetchMatchCandidate matchCandidate,
                                 @Nonnull final Type.Record resultType) {
-        this(indexName, scanParameters, reverse, strictlySorted, Optional.of(matchCandidate), resultType);
+        this(indexName, commonPrimaryKey, scanParameters, useIndexPrefetch, reverse, strictlySorted, Optional.of(matchCandidate), resultType);
     }
 
     private RecordQueryIndexPlan(@Nonnull final String indexName,
+                                 @Nullable final KeyExpression commonPrimaryKey,
                                  @Nonnull final IndexScanParameters scanParameters,
+                                 @Nonnull final RecordQueryPlannerConfiguration.IndexPrefetchUse useIndexPrefetch,
                                  final boolean reverse,
                                  final boolean strictlySorted,
                                  @Nonnull final Optional<? extends ScanWithFetchMatchCandidate> matchCandidateOptional,
                                  @Nonnull final Type resultType) {
         this.indexName = indexName;
+        this.commonPrimaryKey = commonPrimaryKey;
         this.scanParameters = scanParameters;
+        this.useIndexPrefetch = useIndexPrefetch;
         this.reverse = reverse;
         this.strictlySorted = strictlySorted;
         this.matchCandidateOptional = matchCandidateOptional;
         this.resultType = resultType;
+        if (useIndexPrefetch != RecordQueryPlannerConfiguration.IndexPrefetchUse.NONE) {
+            if (commonPrimaryKey == null) {
+                error("Index Prefetch cannot be used without a primary key. Falling back to regular scan.");
+                this.useIndexPrefetch = RecordQueryPlannerConfiguration.IndexPrefetchUse.NONE;
+            }
+            if (scanParameters.getScanType() != IndexScanType.BY_VALUE) {
+                error("Index Prefetch can only be used with VALUE index scan. Falling back to regular scan.");
+                this.useIndexPrefetch = RecordQueryPlannerConfiguration.IndexPrefetchUse.NONE;
+            }
+            if (!FDBDatabaseFactory.instance().getDatabase().getAPIVersion().isAtLeast(APIVersion.API_VERSION_7_1)) {
```

Comment:
```suggestion
            if (!FDBDatabaseFactory.instance().getAPIVersion().isAtLeast(APIVersion.API_VERSION_7_1)) {
```

I think we don't want to call `.getDatabase()` here, because there's not a guarantee that the database with the default cluster file has been created.

Alternatively, I could see the argument that we want to delay this check until `executePlan` so that it can call `store.getRecordContext().isAPIVersionAtLeast(APIVersion.API_VERSION_7_1)`. Right now, these two things should be essentially equivalent because the API version can only be set once globally, but if we adopted the convention that we always checked on the `FDBRecordContext`, then that might set us up better for a future where different databases could have different API versions (or perhaps even different transactions, if we started deriving the API version from the cluster's state, which I think has been proposed).

Reply:
Done. Moved check to executePlan and added another check in the store.